File: c#/methods/BruteForceApp/BruteForce.cs
/*
Resolução de equação não-linear transcendental: método da força bruta.
Programa escrito em c# que implementa os métodos da classe BruteForce.
Autor: Ubiratan Tavares
Data: 07 de novembro de 2024.
*/


using System;
using System.Collections.Generic;

public class BruteForce
{
    private Func<double, double> func;
    private double a;
    private double b;
    private int numPoints;
    private List<(double, double)> roots;

    public BruteForce(Func<double, double> func, double a, double b, int numPoints = 1000)
    {
        this.func = func;
        this.a = a;
        this.b = b;
        this.numPoints = numPoints;
        this.roots = new List<(double, double)>();
    }

    // Método para calcular o passo
    private double CalculateStep()
    {
        return (b - a) / numPoints;
    }

    // Método para verificar se os sinais são opostos
    private bool AreOppositeSigns(double x1, double x2)
    {
        return func(x1) * func(x2) < 0;
    }

    // Método para encontrar raízes usando força bruta
    public List<(double, double)> FindRoots()
    {
        double step = CalculateStep();
        roots.Clear();

        for (int i = 0; i <= numPoints; i++)
        {
            double x = a + i * step;
            if (AreOppositeSigns(x, x + step))
            {
                roots.Add((x, x + step));
            }
        }

        return roots;
    }

    // Método para verificar se a lista está vazia
    private bool IsListEmpty()
    {
        return roots.Count == 0;
    }

    // Método para exibir o resultado
    public void DisplayResult()
    {
        if (IsListEmpty())
        {
            Console.WriteLine("Não encontrou raíz(es) no intervalo.");
        }
        else
        {
            Console.WriteLine("Há raíz(es) no(s) sub-intervalo(s):");
            foreach (var interval in roots)
            {
                Console.WriteLine($"({interval.Item1}, {interval.Item2})");
            }
        }
    }
}

File: matlab/methods/BruteForceApp/BruteForce.m
% BruteForce.m

% Função para calcular o passo
function step = calculate_step(a, b, num_points)
    step = (b - a) / num_points;
end

% Função para verificar se os sinais são opostos
function result = are_opposite_signs(func, a, b)
    result = func(a) * func(b) < 0;
end

% Função de busca forçada para encontrar raízes
function roots = brute_force(func, a, b, num_points)
    roots = [];
    step = calculate_step(a, b, num_points);

    for i = 0:num_points
        x = a + i * step;
        if are_opposite_signs(func, x, x + step)
            roots = [roots; x, x + step];
        end
    end
end

% Função para verificar se a lista está vazia
function result = is_list_empty(roots)
    result = isempty(roots);
end

% Função para exibir o resultado
function display_result(roots)
    if is_list_empty(roots)
        disp("Não encontrou raíz(es) no intervalo.");
    else
        disp("Há raíz(es) no(s) sub-intervalo(s):");
        disp(roots);
    end
end
